### PR TITLE
Add Engine::precompile_compatibility_hash

### LINKED ISSUE #5802: Add `Engine::serialization_compatibility_hash`
#### Feature

Add a new `Engine::serialization_compatibility_hash` method that returns a hash of the "compiler info" appended to precompiled Wasm binaries.

#### Benefit

This would allow precompiled Wasm from multiple configurations/versions of Wasmtime to be safely and efficiently stored together by, for example, keying its storage on `(Hash(wasm), serialization_compat_hash)`.  This may still be a little pessimistic compared with Wasmtime's actual compatibility checks but would be better than the conservative approach of only reusing precompiled binaries on a single host.

#### Implementation

Extract the following section from `append_compiler_info` and reuse it to compute a sha-256 digest of the "compiler info": https://github.com/bytecodealliance/wasmtime/blob/c3c16eb207eccd895f5fbbc4b771bd74ea36d071/crates/wasmtime/src/engine/serialization.rs#L50-L64

#### Alternatives

- There is some overlap between the features this would be used for and Wasmtime's cache system, but the cache system is currently somewhat opaque and difficult to reason about wrt shared storage and the cache "worker"'s performance characteristics.
- This could be done today without changes to Wasmtime by precompiling a dummy module and hashing its `.wasmtime.engine` section, which seems fragile.
- https://github.com/bytecodealliance/wasmtime/issues/3900 suggests splitting out compatibility data into a separate structure, which appears to require a larger effort.

### PR DESCRIPTION
**Pull request by lann**
This method returns a Hash, the output of which can be used to index precompiled binaries from one Engine instance that can be deserialized by another Engine instance.

Closes #5802

### CODE REVIEW

File: crates/wasmtime/src/engine.rs [hunk 1]
```diff
@@ -239,6 +239,22 @@ impl Engine {
...
     }
 
+    /// Returns a value that can be used to check precompiled WebAssembly compatibility.
+    ///
+    /// The outputs of [`Engine::precompile_module`] and [`Engine::precompile_component`]
+    /// are compatible with a different [`Engine`] instance only if the two engines use
+    /// compatible [`Config`]s. The value returned by this function will be the same for two
+    /// [`Engine`]s if a binary from one can be deserialized by the other.
+    #[cfg(compiler)]
+    #[cfg_attr(nightlydoc, doc(cfg(feature = "cranelift")))] // see build.rs
```
Comment (lann) [inline crates/wasmtime/src/engine.rs:249]: Unclear to me if this is needed for this method.

File: crates/wasmtime/src/engine.rs [hunk 2]
```diff
@@ -239,6 +239,22 @@ impl Engine {
...
+    pub fn precompile_compatibility_key(&self) -> impl AsRef<[u8]> {
+        use std::hash::{BuildHasher, Hasher};
+        let data = serialization::build_compiler_info(&self);
+        let hasher = hashbrown::hash_map::DefaultHashBuilder::default().build_hasher();
```
Comment (lann) [inline crates/wasmtime/src/engine.rs:253]: I think a non-crypto hash is safe enough here. If an attacker can control enough of the build_compiler_info to force a collision you probably are already done for.

Comment (bjorn3) [reply crates/wasmtime/src/engine.rs:253]: Hashbrown uses ahash which doesn't guarantee stability between different systems. In fact it returns different values between systems with hardware AES instructions and those that don't. As such using ahash here will break cross-compilation.

Comment (lann) [reply crates/wasmtime/src/engine.rs:253]: Ah good to know, thanks.

Comment (lann) [reply crates/wasmtime/src/engine.rs]: Replaced with sha2, which is probably where I should have started anyway. Its a new dep under some feature sets but it is already used by the cache system.

File: crates/wasmtime/src/engine.rs [hunk 3]
```diff
@@ -239,6 +239,20 @@ impl Engine {
...
+    /// The outputs of [`Engine::precompile_module`] and [`Engine::precompile_component`]
+    /// are compatible with a different [`Engine`] instance only if the two engines use
+    /// compatible [`Config`]s. The value returned by this function will be the same for two
+    /// [`Engine`]s if a binary from one can be deserialized by the other.
+    #[cfg(compiler)]
+    #[cfg_attr(nightlydoc, doc(cfg(feature = "cranelift")))] // see build.rs
+    pub fn precompile_compatibility_key(&self) -> impl AsRef<[u8]> {
+        use sha2::Digest;
+        let data = serialization::build_compiler_info(&self);
+        sha2::Sha256::digest(data)
```
Comment (bjorn3) [inline crates/wasmtime/src/engine.rs:253]: sha512 is faster on 64bit systems as it can consume more bytes at once. Not that it really matters here. This hashing is highly unlikely to ever become a bottleneck.

File: crates/wasmtime/src/engine.rs [hunk 4]
```diff
@@ -239,6 +239,20 @@ impl Engine {
...
+    ///
+    /// The outputs of [`Engine::precompile_module`] and [`Engine::precompile_component`]
+    /// are compatible with a different [`Engine`] instance only if the two engines use
+    /// compatible [`Config`]s. The value returned by this function will be the same for two
+    /// [`Engine`]s if a binary from one can be deserialized by the other.
+    #[cfg(compiler)]
+    #[cfg_attr(nightlydoc, doc(cfg(feature = "cranelift")))] // see build.rs
+    pub fn precompile_compatibility_key(&self) -> impl AsRef<[u8]> {
+        use sha2::Digest;
+        let data = serialization::build_compiler_info(&self);
```
Comment (alexcrichton) [inline crates/wasmtime/src/engine.rs:252]: I think it's fine to just return the whole blob here and let embedders hash it if they want. That avoids a dependency here and should enable a bit more flexibility for embedders as well.

File: crates/wasmtime/src/engine.rs [hunk 5]
```diff
@@ -239,6 +239,20 @@ impl Engine {
         Ok(mmap.to_vec())
     }
 
+    /// Returns a value that can be used to check precompiled WebAssembly compatibility.
+    ///
+    /// The outputs of [`Engine::precompile_module`] and [`Engine::precompile_component`]
+    /// are compatible with a different [`Engine`] instance only if the two engines use
+    /// compatible [`Config`]s. The value returned by this function will be the same for two
+    /// [`Engine`]s if a binary from one can be deserialized by the other.
```
Comment (alexcrichton) [inline crates/wasmtime/src/engine.rs:247]: I think the documentation here may want to be a bit more nuanced because if two blobs are the same then you should be able to load between engines but if they're different they may still load correctly. For example different values of `generate_address_map` will produce different blobs but that's not checked when loading modules into engines.

I think it may be best to reword this as something along the lines of "if two blobs are equal artifacts from one engine are guaranteed to be loadable in another" or something like that.

File: crates/wasmtime/src/engine.rs (hunk 1 above)
Comment (alexcrichton) [reply crates/wasmtime/src/engine.rs:249]: If this compiles without the `#[cfg]` no need for the attribute. If the `#[cfg]` is required, though, then this attribute is also required for better display in documentation.

File: crates/wasmtime/src/engine.rs (hunk 5 above)
Comment (lann) [reply crates/wasmtime/src/engine.rs:247]: Yeah that's what I was trying to convey but "if" is ambiguous. I'll be more explicit about it.

**Comment by lann** [comment]
Finally getting back around to this, I realized that the cache system already generates a key like this, and sure enough this has exactly what I was looking for: https://github.com/bytecodealliance/wasmtime/blob/83d00fea4a881ab1160bf0a609ac463643bc0821/crates/wasmtime/src/module.rs#L1141-L1151

Reusing this code as-is would require changing the function signature a bit, e.g.:

```rust
pub fn precompile_compatibility_hash(&self) -> impl std::hash::Hash {
    crate::module::HashedEngineCompileEnv(self)
}
```

**Comment by alexcrichton** [comment]
Sounds reasonable to me!

**Review by jameysharp — APPROVED** [review]
This looks sensible to me and appears to be exactly what Alex already said sounded good, so I'll go ahead and merge it. Thank you!
